Add owner reference to update-check job so that it points to Mattermost resource

## Description

When modifying `spec.image` or `spec.version` of Mattermost resource the update-check job is run to ensure image is correct.
However, the job is not owned by the Mattermost resource which causes Operator to not to pick up when the job completes and carry on with update.

### Possible solution

Very simple and reliable solution would be to add `OwnerReference` to the job, so that it is owned by Mattermost resource.
We already have [function](https://github.com/mattermost/mattermost-operator/blob/272928435a67a730fe50583fee309a738c9d8536/pkg/resources/create_resources.go#L48) that adds owner references when creating objects. We could use it [when lunching the update job](https://github.com/mattermost/mattermost-operator/blob/272928435a67a730fe50583fee309a738c9d8536/pkg/resources/update_job.go#L19)

@Szymongib When setting up this repo locally and running command `make build-image kind-load-image deploy` I get following error:

```
go: added sigs.k8s.io/yaml v1.2.0
mv: rename /Users/shubhamkumar/go/src/github.com/mattermost/mattermost-operator/bin/controller-gen to /Users/shubhamkumar/go/src/github.com/mattermost/mattermost-operator/bin/controller-gen-v0.7.0: No such file or directory
make: *** [/Users/shubhamkumar/go/src/github.com/mattermost/mattermost-operator/bin/controller-gen] Error 1
```

@Szymongib Thanks. I am getting following now when running `make build-image kind-load-image deploy`:

```
namespace/mattermost-operator configured
cd config/manager && /home/lucifer/go/bin/kustomize edit set image mattermost-operator="mattermost/mattermost-operator:test"
/bin/sh: line 1: /home/lucifer/go/bin/kustomize: No such file or directory
make: *** [Makefile:153: deploy] Error 127
```

--- a/controllers/mattermost/clusterinstallation/mattermost.go
+++ b/controllers/mattermost/clusterinstallation/mattermost.go
@@ -419,7 +419,7 @@ func (r *ClusterInstallationReconciler) checkUpdateJob(
 	}
 	if !isSameImage {
 		reqLogger.Info("Mattermost image changed, restarting update job")
-		err := r.Resources.RestartMattermostUpdateJob(job, desired)
+		err := r.Resources.RestartMattermostUpdateJob(mattermost, job, desired, reqLogger)
 		if err != nil {
 			return nil, errors.Wrap(err, "failed to restart update job")
 		}

--- a/controllers/mattermost/clusterinstallation/mattermost.go
+++ b/controllers/mattermost/clusterinstallation/mattermost.go
@@ -20,6 +20,7 @@ import (
 	k8sClient "sigs.k8s.io/controller-runtime/pkg/client"
 
 	mattermostv1alpha1 "github.com/mattermost/mattermost-operator/apis/mattermost/v1alpha1"
+	"github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
 	"github.com/mattermost/mattermost-operator/pkg/database"
 	mattermostApp "github.com/mattermost/mattermost-operator/pkg/mattermost"
 )
@@ -346,6 +347,7 @@ func (r *ClusterInstallationReconciler) isMainContainerImageSame(
 // updateMattermostDeployment performs deployment update if necessary.
 // If a deployment update is necessary, an update job is launched to check new image.
 func (r *ClusterInstallationReconciler) updateMattermostDeployment(
+	owner *v1beta1.Mattermost,
 	mattermost *mattermostv1alpha1.ClusterInstallation,
 	current *appsv1.Deployment,
 	desired *appsv1.Deployment,
@@ -370,7 +372,7 @@ func (r *ClusterInstallationReconciler) updateMattermostDeployment(
 
 	reqLogger.Info("Current image is not the same as the requested, will upgrade the Mattermost installation")
 
-	job, err := r.checkUpdateJob(mattermost, desired, reqLogger)
+	job, err := r.checkUpdateJob(owner, mattermost, desired, reqLogger)
 	if job != nil {
 		// Job is done, need to cleanup
 		defer r.cleanupUpdateJob(job, reqLogger)
@@ -386,6 +388,7 @@ func (r *ClusterInstallationReconciler) updateMattermostDeployment(
 
 // checkUpdateJob checks whether update job status. In case job is not running it is launched
 func (r *ClusterInstallationReconciler) checkUpdateJob(
+	owner *v1beta1.Mattermost,
 	mattermost *mattermostv1alpha1.ClusterInstallation,
 	desired *appsv1.Deployment,
 	reqLogger logr.Logger,
@@ -397,7 +400,7 @@ func (r *ClusterInstallationReconciler) checkUpdateJob(
 		if k8sErrors.IsNotFound(err) {
 			// Job is not running, let's launch
 			reqLogger.Info("Launching update image job")
-			if err = r.Resources.LaunchMattermostUpdateJob(mattermost.Namespace, desired); err != nil {
+			if err = r.Resources.LaunchMattermostUpdateJob(owner, mattermost.Namespace, desired, reqLogger); err != nil {
 				return nil, errors.Wrap(err, "Launching update image job failed")
 			}
 			return nil, errors.New("Began update image job")

--- a/controllers/mattermost/clusterinstallation/mattermost.go
+++ b/controllers/mattermost/clusterinstallation/mattermost.go
@@ -20,7 +20,6 @@ import (
 	k8sClient "sigs.k8s.io/controller-runtime/pkg/client"
 
 	mattermostv1alpha1 "github.com/mattermost/mattermost-operator/apis/mattermost/v1alpha1"
-	"github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
 	"github.com/mattermost/mattermost-operator/pkg/database"
 	mattermostApp "github.com/mattermost/mattermost-operator/pkg/mattermost"
 )
@@ -347,7 +346,6 @@ func (r *ClusterInstallationReconciler) isMainContainerImageSame(
 // updateMattermostDeployment performs deployment update if necessary.
 // If a deployment update is necessary, an update job is launched to check new image.
 func (r *ClusterInstallationReconciler) updateMattermostDeployment(
-	owner *v1beta1.Mattermost,
 	mattermost *mattermostv1alpha1.ClusterInstallation,
 	current *appsv1.Deployment,
 	desired *appsv1.Deployment,
@@ -372,7 +370,7 @@ func (r *ClusterInstallationReconciler) updateMattermostDeployment(
 
 	reqLogger.Info("Current image is not the same as the requested, will upgrade the Mattermost installation")
 
-	job, err := r.checkUpdateJob(owner, mattermost, desired, reqLogger)
+	job, err := r.checkUpdateJob(mattermost, desired, reqLogger)
 	if job != nil {
 		// Job is done, need to cleanup
 		defer r.cleanupUpdateJob(job, reqLogger)
@@ -388,7 +386,6 @@ func (r *ClusterInstallationReconciler) updateMattermostDeployment(
 
 // checkUpdateJob checks whether update job status. In case job is not running it is launched
 func (r *ClusterInstallationReconciler) checkUpdateJob(
-	owner *v1beta1.Mattermost,
 	mattermost *mattermostv1alpha1.ClusterInstallation,
 	desired *appsv1.Deployment,
 	reqLogger logr.Logger,
@@ -400,7 +397,7 @@ func (r *ClusterInstallationReconciler) checkUpdateJob(
 		if k8sErrors.IsNotFound(err) {
 			// Job is not running, let's launch
 			reqLogger.Info("Launching update image job")
-			if err = r.Resources.LaunchMattermostUpdateJob(owner, mattermost.Namespace, desired, reqLogger); err != nil {
+			if err = r.Resources.LaunchMattermostUpdateJob(mattermost, mattermost.Namespace, desired, reqLogger); err != nil {
 				return nil, errors.Wrap(err, "Launching update image job failed")
 			}
 			return nil, errors.New("Began update image job")

--- a/controllers/mattermost/mattermost/mattermost.go
+++ b/controllers/mattermost/mattermost/mattermost.go
@@ -377,7 +377,7 @@ func (r *MattermostReconciler) checkUpdateJob(
 	}
 	if !isSameImage {
 		reqLogger.Info("Mattermost image changed, restarting update job")
-		err := r.Resources.RestartMattermostUpdateJob(job, baseDeployment)
+		err := r.Resources.RestartMattermostUpdateJob(mattermost, job, baseDeployment, reqLogger)
 		if err != nil {
 			return nil, errors.Wrap(err, "failed to restart update job")
 		}

--- a/controllers/mattermost/mattermost/mattermost.go
+++ b/controllers/mattermost/mattermost/mattermost.go
@@ -307,7 +307,6 @@ func (r *MattermostReconciler) isMainContainerImageSame(
 // updateMattermostDeployment performs deployment update if necessary.
 // If a deployment update is necessary, an update job is launched to check new image.
 func (r *MattermostReconciler) updateMattermostDeployment(
-	owner *v1beta1.Mattermost,
 	mattermost *mmv1beta.Mattermost,
 	current *appsv1.Deployment,
 	desired *appsv1.Deployment,
@@ -331,7 +330,7 @@ func (r *MattermostReconciler) updateMattermostDeployment(
 
 	reqLogger.Info("Current image is not the same as the requested, will upgrade the Mattermost installation")
 
-	job, err := r.checkUpdateJob(owner, mattermost.Namespace, desired, reqLogger)
+	job, err := r.checkUpdateJob(mattermost, mattermost.Namespace, desired, reqLogger)
 	if job != nil {
 		// Job is done, need to cleanup
 		defer r.cleanupUpdateJob(job, reqLogger)
@@ -347,7 +346,7 @@ func (r *MattermostReconciler) updateMattermostDeployment(
 
 // checkUpdateJob checks whether update job status. In case job is not running it is launched
 func (r *MattermostReconciler) checkUpdateJob(
-	owner *v1beta1.Mattermost,
+	mattermost *v1beta1.Mattermost,
 	jobNamespace string,
 	baseDeployment *appsv1.Deployment,
 	reqLogger logr.Logger,
@@ -357,7 +356,7 @@ func (r *MattermostReconciler) checkUpdateJob(
 	if err != nil {
 		if k8sErrors.IsNotFound(err) {
 			reqLogger.Info("Launching update image job")
-			if err = r.Resources.LaunchMattermostUpdateJob(owner, jobNamespace, baseDeployment, reqLogger); err != nil {
+			if err = r.Resources.LaunchMattermostUpdateJob(mattermost, jobNamespace, baseDeployment, reqLogger); err != nil {
 				return nil, errors.Wrap(err, "Launching update image job failed")
 			}
 			return nil, errors.New("Began update image job")

--- a/controllers/mattermost/mattermost/mattermost.go
+++ b/controllers/mattermost/mattermost/mattermost.go
@@ -3,9 +3,11 @@ package mattermost
 import (
 	"context"
 	"fmt"
+
 	"github.com/mattermost/mattermost-operator/pkg/resources"
 
 	"github.com/go-logr/logr"
+	"github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
 	mmv1beta "github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
 	mattermostApp "github.com/mattermost/mattermost-operator/pkg/mattermost"
 	"github.com/pkg/errors"
@@ -305,6 +307,7 @@ func (r *MattermostReconciler) isMainContainerImageSame(
 // updateMattermostDeployment performs deployment update if necessary.
 // If a deployment update is necessary, an update job is launched to check new image.
 func (r *MattermostReconciler) updateMattermostDeployment(
+	owner *v1beta1.Mattermost,
 	mattermost *mmv1beta.Mattermost,
 	current *appsv1.Deployment,
 	desired *appsv1.Deployment,
@@ -328,7 +331,7 @@ func (r *MattermostReconciler) updateMattermostDeployment(
 
 	reqLogger.Info("Current image is not the same as the requested, will upgrade the Mattermost installation")
 
-	job, err := r.checkUpdateJob(mattermost.Namespace, desired, reqLogger)
+	job, err := r.checkUpdateJob(owner, mattermost.Namespace, desired, reqLogger)
 	if job != nil {
 		// Job is done, need to cleanup
 		defer r.cleanupUpdateJob(job, reqLogger)
@@ -344,6 +347,7 @@ func (r *MattermostReconciler) updateMattermostDeployment(
 
 // checkUpdateJob checks whether update job status. In case job is not running it is launched
 func (r *MattermostReconciler) checkUpdateJob(
+	owner *v1beta1.Mattermost,
 	jobNamespace string,
 	baseDeployment *appsv1.Deployment,
 	reqLogger logr.Logger,
@@ -353,7 +357,7 @@ func (r *MattermostReconciler) checkUpdateJob(
 	if err != nil {
 		if k8sErrors.IsNotFound(err) {
 			reqLogger.Info("Launching update image job")
-			if err = r.Resources.LaunchMattermostUpdateJob(jobNamespace, baseDeployment); err != nil {
+			if err = r.Resources.LaunchMattermostUpdateJob(owner, jobNamespace, baseDeployment, reqLogger); err != nil {
 				return nil, errors.Wrap(err, "Launching update image job failed")
 			}
 			return nil, errors.New("Began update image job")

--- a/controllers/mattermost/mattermost/mattermost_test.go
+++ b/controllers/mattermost/mattermost/mattermost_test.go
@@ -278,6 +278,8 @@ func TestCheckMattermost(t *testing.T) {
 		// Assert owner references are set here
 		assert.NotEmpty(t, job.ObjectMeta.OwnerReferences)
 		assert.Equal(t, "Mattermost", job.ObjectMeta.OwnerReferences[0].Kind)
+		assert.Equal(t, mmName, job.ObjectMeta.OwnerReferences[0].Name)
+		assert.Equal(t, "installation.mattermost.com/v1beta1", job.ObjectMeta.OwnerReferences[0].APIVersion)
 
 		// Set job status to succeeded so that test can proceed
 		now := metav1.Now()

--- a/controllers/mattermost/mattermost/mattermost_test.go
+++ b/controllers/mattermost/mattermost/mattermost_test.go
@@ -3,12 +3,11 @@ package mattermost
 import (
 	"context"
 	"fmt"
-	"testing"
-
 	"github.com/go-logr/logr"
 	"github.com/sirupsen/logrus"
 	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"testing"
 
 	"github.com/mattermost/mattermost-operator/pkg/resources"
 
@@ -237,27 +236,7 @@ func TestCheckMattermost(t *testing.T) {
 	})
 
 	t.Run("deployment", func(t *testing.T) {
-		updateName := "mattermost-update-check"
-		now := metav1.Now()
-		job := &batchv1.Job{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      updateName,
-				Namespace: mm.GetNamespace(),
-			},
-			Spec: batchv1.JobSpec{Template: corev1.PodTemplateSpec{
-				Spec: corev1.PodSpec{
-					Containers: []corev1.Container{
-						{Name: mmv1beta.MattermostAppContainerName, Image: mm.GetImageName()},
-					},
-				},
-			}},
-			Status: batchv1.JobStatus{
-				Succeeded:      1,
-				CompletionTime: &now,
-			},
-		}
-		err = reconciler.Client.Create(context.TODO(), job)
-		require.NoError(t, err)
+		updateJobName := "mattermost-update-check"
 
 		err = reconciler.checkMattermostDeployment(mm, dbInfo, fileStoreInfo, currentMMStatus, logger)
 		assert.NoError(t, err)
@@ -285,8 +264,31 @@ func TestCheckMattermost(t *testing.T) {
 
 		err = reconciler.Client.Update(context.TODO(), modified)
 		require.NoError(t, err)
+
+		// Update job should be launched, we expect error
+		err = reconciler.checkMattermostDeployment(mm, dbInfo, fileStoreInfo, currentMMStatus, logger)
+		require.Error(t, err)
+
+		// Assert the job was created
+		job := &batchv1.Job{}
+		err = reconciler.Client.Get(context.TODO(), types.NamespacedName{Name: updateJobName, Namespace: mmNamespace}, job)
+		require.NoError(t, err)
+
+		// TODO: assert owner references are set here
+
+		// Set job status to succeeded so that test can proceed
+		now := metav1.Now()
+		job.Status =  batchv1.JobStatus{
+			Succeeded:      1,
+			CompletionTime: &now,
+		}
+		err = reconciler.Client.Update(context.TODO(), job)
+		require.NoError(t, err)
+
+		// Job is marked as succeeded, should proceed now.
 		err = reconciler.checkMattermostDeployment(mm, dbInfo, fileStoreInfo, currentMMStatus, logger)
 		require.NoError(t, err)
+
 		err = reconciler.Client.Get(context.TODO(), types.NamespacedName{Name: mmName, Namespace: mmNamespace}, found)
 		require.NoError(t, err)
 		assert.Equal(t, original.GetLabels(), found.GetLabels())

--- a/controllers/mattermost/mattermost/mattermost_test.go
+++ b/controllers/mattermost/mattermost/mattermost_test.go
@@ -3,11 +3,12 @@ package mattermost
 import (
 	"context"
 	"fmt"
+	"testing"
+
 	"github.com/go-logr/logr"
 	"github.com/sirupsen/logrus"
 	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"testing"
 
 	"github.com/mattermost/mattermost-operator/pkg/resources"
 
@@ -274,11 +275,13 @@ func TestCheckMattermost(t *testing.T) {
 		err = reconciler.Client.Get(context.TODO(), types.NamespacedName{Name: updateJobName, Namespace: mmNamespace}, job)
 		require.NoError(t, err)
 
-		// TODO: assert owner references are set here
+		// Assert owner references are set here
+		assert.NotEmpty(t, job.ObjectMeta.OwnerReferences)
+		assert.Equal(t, "Mattermost", job.ObjectMeta.OwnerReferences[0].Kind)
 
 		// Set job status to succeeded so that test can proceed
 		now := metav1.Now()
-		job.Status =  batchv1.JobStatus{
+		job.Status = batchv1.JobStatus{
 			Succeeded:      1,
 			CompletionTime: &now,
 		}

--- a/pkg/resources/create_resources.go
+++ b/pkg/resources/create_resources.go
@@ -52,12 +52,13 @@ func (r *ResourceHelper) Create(owner v1.Object, desired Object, reqLogger logr.
 	if err != nil {
 		return errors.Wrap(err, "failed to apply annotation to the resource")
 	}
-	err = r.client.Create(context.TODO(), desired)
+
+	err = controllerutil.SetControllerReference(owner, desired, r.scheme)
 	if err != nil {
-		return errors.Wrap(err, "failed to create resource")
+		return errors.Wrap(err, "failed to set owner reference")
 	}
 
-	return controllerutil.SetControllerReference(owner, desired, r.scheme)
+	return r.client.Create(context.TODO(), desired)
 }
 
 func (r *ResourceHelper) Update(current, desired Object, reqLogger logr.Logger) error {

--- a/pkg/resources/update_job.go
+++ b/pkg/resources/update_job.go
@@ -4,7 +4,6 @@ import (
 	"context"
 
 	"github.com/go-logr/logr"
-	"github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
 	mattermostApp "github.com/mattermost/mattermost-operator/pkg/mattermost"
 	"github.com/pkg/errors"
 	appsv1 "k8s.io/api/apps/v1"
@@ -19,7 +18,7 @@ import (
 const UpdateJobName = "mattermost-update-check"
 
 func (r *ResourceHelper) LaunchMattermostUpdateJob(
-	owner *v1beta1.Mattermost,
+	owner metav1.Object,
 	jobNamespace string,
 	baseDeployment *appsv1.Deployment,
 	reqLogger logr.Logger,

--- a/pkg/resources/update_job.go
+++ b/pkg/resources/update_job.go
@@ -3,6 +3,8 @@ package resources
 import (
 	"context"
 
+	"github.com/go-logr/logr"
+	"github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
 	mattermostApp "github.com/mattermost/mattermost-operator/pkg/mattermost"
 	"github.com/pkg/errors"
 	appsv1 "k8s.io/api/apps/v1"
@@ -17,12 +19,14 @@ import (
 const UpdateJobName = "mattermost-update-check"
 
 func (r *ResourceHelper) LaunchMattermostUpdateJob(
+	owner *v1beta1.Mattermost,
 	jobNamespace string,
 	baseDeployment *appsv1.Deployment,
+	reqLogger logr.Logger,
 ) error {
 	job := PrepareMattermostJobTemplate(UpdateJobName, jobNamespace, baseDeployment)
 
-	err := r.client.Create(context.TODO(), job)
+	err := r.Create(owner, job, reqLogger)
 	if err != nil && !k8sErrors.IsAlreadyExists(err) {
 		return err
 	}

--- a/pkg/resources/update_job.go
+++ b/pkg/resources/update_job.go
@@ -35,8 +35,10 @@ func (r *ResourceHelper) LaunchMattermostUpdateJob(
 
 // RestartMattermostUpdateJob removes existing update job if it exists and creates new one.
 func (r *ResourceHelper) RestartMattermostUpdateJob(
+	owner metav1.Object,
 	currentJob *batchv1.Job,
 	deployment *appsv1.Deployment,
+	reqLogger logr.Logger,
 ) error {
 	err := r.client.Delete(context.TODO(), currentJob, k8sClient.PropagationPolicy(metav1.DeletePropagationBackground))
 	if err != nil && !k8sErrors.IsNotFound(err) {
@@ -45,7 +47,7 @@ func (r *ResourceHelper) RestartMattermostUpdateJob(
 
 	job := PrepareMattermostJobTemplate(UpdateJobName, currentJob.Namespace, deployment)
 
-	err = r.client.Create(context.TODO(), job)
+	err = r.Create(owner, job, reqLogger)
 	if err != nil {
 		return err
 	}

